src/backend/distributed/planner/distributed_planner.c

```diff
@@ -457,19 +457,12 @@ IsMultiTaskPlan(DistributedPlan *distributedPlan)
 
 
 /*
- * IsUpdateOrDelete returns true if the query performs update or delete.
+ * IsUpdateOrDelete returns true if the query performs an update or delete.
  */
 bool
 IsUpdateOrDelete(DistributedPlan *distributedPlan)
 {
-	CmdType commandType = distributedPlan->operation;
-
-	if (commandType == CMD_UPDATE || commandType == CMD_DELETE)
-	{
-		return true;
-	}
-
-	return false;
+	return distributedPlan->modLevel >= ROW_MODIFY_NONCOMMUTATIVE;
```

Comment:
I think the only caller of this function is `IsMultiShardModifyPlan`, which is quite specifically looking for multi-shard UPDATE/DELETE and not multi-shard upserts. It might be ok to continue checking command type for now, at least until we can push down CTEs in multi-shard statements.

For context: `IsMultiShardModifyPlan` function is called in `CreateDistributedPlannedStmt` to avoid planning prepared update/delete command that has a `distribution column = $1` filter as a multi-shard update/delete when it could be a single-shard update/delete if the parameter was resolved. For multi-shard upserts we can tolerate parameters in the distribution column since we can defer shard pruning to the executor, so you could see a slight performance improvement when using a prepared statement.